File: src/model/domain/model.service.ts
import { Injectable } from '@nestjs/common';
import { ModelId, Model, TransitionId, Transition } from './model';

@Injectable()
export class ModelService {
    models: Map<ModelId, Model>

    saveModel(model: Model) {
        this.models.set(model.id, model);
    }

    findModel(modelId: ModelId): Model {
        const model = this.models.get(modelId);
        if (!model) {
            throw Error(`Model ${modelId} not found`);
        }
        return model;
    }
}

File: execution-service/model/model_service.go
package model

import (
	"execution-service/domain"
	"fmt"
)

type ModelService struct {
	models    map[domain.ModelId]domain.Model
	modelPort ModelPort
}

func NewModelService(modelPort ModelPort) ModelService {
	return ModelService{
		models:    make(map[string]domain.Model),
		modelPort: modelPort,
	}
}

func (service *ModelService) SaveModel(model domain.Model) {
	service.models[model.Id] = model
}

func (service *ModelService) FindModelById(modelId domain.ModelId) (domain.Model, error) {
	model, exists := service.models[modelId]
	if exists {
		return model, nil
	}
	model, err := service.modelPort.FindModelById(modelId)
	if err != nil {
		return domain.Model{}, fmt.Errorf("model %s not found in model-service", modelId)
	}
	model.ComputeHash()
	service.SaveModel(model)
	return model, nil
}
